### PR TITLE
[KIE-1492] Allow KieRuntimeBuilder to also create and provide StatelessKieSession

### LINKED ISSUE #1492: Allow KieRuntimeBuilder to also create and provide StatelessKieSession
At the moment the [`KieRuntimeBuilder`](https://github.com/apache/incubator-kie-drools/blob/main/kie-api/src/main/java/org/kie/api/runtime/KieRuntimeBuilder.java) only allow to retrieve `KieSession` s, but not `StatelessKieSession` s. Even worse when a KieSession configured as stateless is required through the current API it creates a stateful one anyway, making things even more confused.

Finally the impossibility of creating a `StatelessKieSession` from `KieRuntimeBuilder` is also the indirect cause of a memory leak with the Kogito integration as reported [here](https://github.com/apache/incubator-kie-issues/issues/1491).

### PR DESCRIPTION
**Pull request by mariofusco**
Closes: https://github.com/apache/incubator-kie-issues/issues/1492

### CODE REVIEW

File: kie-api/src/main/java/org/kie/api/runtime/RuntimeSession.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.kie.api.runtime;
+
+import org.kie.api.KieBase;
+
+import java.util.Map;
+
+public interface RuntimeSession extends CommandExecutor {
```
Comment (mariofusco) [inline kie-api/src/main/java/org/kie/api/runtime/RuntimeSession.java:25]: I introduced this interface to have a usable common ground between `KieSession` and `StatelessKieSession`. Note that they still exists with the same methods as before, so this change is totally backward compatible.

File: kie-dmn/kie-dmn-core-jsr223/pom.xml [hunk 2]
```diff
@@ -112,21 +112,10 @@
       <artifactId>kie-dmn-core-jsr223-jq</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.openjdk.nashorn</groupId>
+      <artifactId>nashorn-core</artifactId>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
-  <profiles>
```
Comment (mariofusco) [inline kie-dmn/kie-dmn-core-jsr223/pom.xml:116]: These and other changes to poms file are actually not related to this fix, but they are meant to fix some tech debt to support old JVM versions that we don't need anymore.

File: drools-quarkus-extension/drools-quarkus-integration-test/src/test/java/org/drools/quarkus/test/RuntimeTest.java [hunk 3]
```diff
@@ -42,30 +47,59 @@ public class RuntimeTest {
...
+    public void testStatelessDTableEvaluation() {
+        testSimpleDrl(runtimeBuilder.newStatelessKieSession("statelessCanDrinkKSDTable"), "org.drools.dtable", false);
+    }
+
+    @Test
+    public void testStatelessYamlEvaluation() {
+        testSimpleDrl(runtimeBuilder.newStatelessKieSession("statelessCanDrinkKSYaml"), "org.drools.yaml", false);
+    }
+
+    private void testSimpleDrl(RuntimeSession session, String assetPackage, boolean stateful) {
```
Comment (mariofusco) [inline drools-quarkus-extension/drools-quarkus-integration-test/src/test/java/org/drools/quarkus/test/RuntimeTest.java:79]: Here I'm using the new `RuntimeSession` interface to test both `KieSession` and `StatelessKieSession` with the same method.

File: kie-api/src/main/java/org/kie/api/command/KieCommands.java [hunk 4]
```diff
@@ -115,6 +117,10 @@ Command newQuery(String identifier,
                      String name,
                      Object[] arguments);
 
+    default BatchExecutionCommand newBatchExecution(Command... commands) {
```
Comment (gitgabrio) [inline kie-api/src/main/java/org/kie/api/command/KieCommands.java:120]: HI @mariofusco 
is this related to the overall task ? If so, could. you please explain why ?

File: kie-api/src/main/java/org/kie/api/runtime/StatelessKieSession.java [hunk 5]
```diff
@@ -106,51 +103,6 @@
...
-     *
-     * @param name
-     */
-    void unregisterChannel(String name);
-
-    /**
-     * @return a map with all registered channels.
-     */
-    Map<String, Channel> getChannels();
+public interface StatelessKieSession extends StatelessRuleSession, StatelessProcessSession, RuntimeSession, KieRuntimeEventManager {
```
Comment (gitgabrio) [inline kie-api/src/main/java/org/kie/api/runtime/StatelessKieSession.java:106]: I did not notice that before - but at very first sight I have the impression the hierarchies are somehow turned upside:
StatelessRuleSession and StatelessRuleSession should be "specialization" of StatelessKieSession, but as it is, StatelessKieSession is an aggregation of the two, creating an hard-binding: wdyt ?

**Review by gitgabrio — COMMENTED** [review]
Thanks @mariofusco !!!
For the moment being I left some minor comments.
As side-note, and not specifically related to this PR, I have the impression that the whole "KieSession" (statefull/stateless/etc) hierarchy would benefit from a good review, but surely not in this PR
I need slighlty more time to dig in the code itself

File: kie-api/src/main/java/org/kie/api/command/KieCommands.java (hunk 4 above)
Comment (mariofusco) [reply kie-api/src/main/java/org/kie/api/command/KieCommands.java:120]: Not strictly related to this fix, it's just a convenience method that I added to create batch commands in a single line that I used in my test.

File: kie-api/src/main/java/org/kie/api/runtime/StatelessKieSession.java (hunk 5 above)
Comment (mariofusco) [reply kie-api/src/main/java/org/kie/api/runtime/StatelessKieSession.java:106]: Unfortunately you're right, or at least the name of those interfaces are quite misleading, but we cannot change them if we want to keep backward compatibility.

**Comment by tkobayas** [comment]
Thanks, Mario.

GHA kogito-runtimes (and other downstreams):
```
2024-09-25T16:38:02.4742156Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project jbpm-flow: Compilation failure
2024-09-25T16:38:02.4744813Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/jbpm/jbpm-flow/src/main/java/org/jbpm/process/instance/DummyKnowledgeRuntime.java:[65,1] org.jbpm.process.instance.DummyKnowledgeRuntime is not abstract and does not override abstract method <T>execute(org.kie.api.command.Command<T>) in org.kie.api.runtime.CommandExecutor
```
`DummyKnowledgeRuntime` needs to implement `execute` or `CommandExecutor.execute` should have a default implementation.

**Comment by mariofusco** [comment]
> Thanks, Mario.
> 
> GHA kogito-runtimes (and other downstreams):
> 
> ```
> 2024-09-25T16:38:02.4742156Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project jbpm-flow: Compilation failure
> 2024-09-25T16:38:02.4744813Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/jbpm/jbpm-flow/src/main/java/org/jbpm/process/instance/DummyKnowledgeRuntime.java:[65,1] org.jbpm.process.instance.DummyKnowledgeRuntime is not abstract and does not override abstract method <T>execute(org.kie.api.command.Command<T>) in org.kie.api.runtime.CommandExecutor
> ```
> 
> `DummyKnowledgeRuntime` needs to implement `execute` or `CommandExecutor.execute` should have a default implementation.

This is the result of a **HUGE** technical debt caused by the tight coupling between Drools and Kogito's processes that has no longer reason to exist. I spent a couple of hours trying to refactor Kogito and lower this coupling, but unfortunately this is clearly a multi-days effort, so I gave up for now and fixed the issue in the most straightforward way with this pull request https://github.com/apache/incubator-kie-kogito-runtimes/pull/3681

That said, I'm still willing to solve this debt with more calm when I will have time to dedicate to this, possibly with the help of @elguardian and @fjtirado. I already have a very preliminary question for you 2: is that `DummyKnowledgeRuntime` still necessary for some reasons (and why) or could it be totally removed? I'm asking because this seems to be a hack that it is also a source of a big part of the issues that I found during this first refactor attempt.

@tkobayas @gitgabrio please review again.

**Comment by fjtirado** [comment]
Hi Mario,
DummyKnowledgeRuntime should be a goner for good.
Besides that, I think that  the final result of the refactor (which I agree
is a lengthy task) should be that the process engine (which is currently
used by BPMN and Sonataflow) no longer depends on drools*.
If we agree that's the goal, I believe it is better for you to work on that
alone (meaning quietly, without continuous interference by me ;)). Once you
have the PR, let's review it also quietly.


* The dependency with drools should be in the RuleNode, which is the part
of the engine related with rules.

**Comment by gitgabrio** [comment]
@mariofusco @fjtirado @elguardian 
There is an ongoing task for that: I started make the analysis to remove all the *processes references from drools repository, but I have to finish another ticket before.
@elguardian ? 
Anyway, let's avoid duplicate the effort, and stay sync with that

**Review by tkobayas — APPROVED** [review]

**Comment by mariofusco** [comment]
> @mariofusco @fjtirado @elguardian There is an ongoing task for that: I started make the analysis to remove all the *processes references from drools repository, but I have to finish another ticket before. @elguardian ? Anyway, let's avoid duplicate the effort, and stay sync with that

I agree on avoid duplicating the effort, and if possible I would like to work on this on small-ish steps. For now I will try to remove that `DummyKnowledgeRuntime` class as suggested by @fjtirado or is there anybody else working on this? /cc @elguardian 

@gitgabrio If this pull request ok for you? If so can you please approve so I could merge and move forward with that rework?

**Comment by mariofusco** [comment]
I just realized there are a few failing tests, another side effect of that coupling I guess :(
I will fix them immediately.

**Comment by gitgabrio** [comment]
@mariofusco 

> If this pull request ok for you? 

Fine for me as soon as the build is green 👍 

> For now I will try to remove that DummyKnowledgeRuntime class as suggested by @fjtirado or is there anybody else working on this? /cc @elguardian

Fine for me
